src/shim/shim.c

```diff
@@ -15,43 +16,101 @@
 #include "shim/shim_logger.h"
 #include "support/logger/logger.h"
 
-// Whether Shadow is using INTERPOSE_PRELOAD
-static bool _using_interpose_preload;
+typedef enum {
+    INTERPOSE_NONE,
+    INTERPOSE_PRELOAD,
+    INTERPOSE_PTRACE,
+    INTERPOSE_PTRACE_NOIPC,
+} InterposeType;
```

Comment:
Done-ish.

Here in shim.c, with the benefit of hindsight I think it's easier to have two independent variables tracking whether we're using ptrace and whether we're using preload ipc.

I changed some other places based on your suggestion here.

I left the command-line option as "ptrace" and class ThreadPtrace to get "preload+ptrace", mostly so that I could flip the default to use preload+ptrace in this PR (thus validating it in the tests) without also bringing in a new big rename. If we do want to change those, I suggest we do it in a follow-up PR as a pure refactor.

Reply:
Makes sense!